packages/react/src/List/OrderedList.tsx

```diff
@@ -0,0 +1,25 @@
+import React, {type Ref} from 'react'
+import clsx from 'clsx'
+import type {BaseProps} from '../component-helpers'
+import {ListContext} from './UnorderedList'
+
+import styles from './List.module.css'
+
+export type OrderedListProps = BaseProps<HTMLElement> & {
+  /**
+   * React.ReactNode and React.ReactNode[] are valid children.
+   */
+  children?: React.ReactNode | React.ReactNode[]
+  /**
+   * Ref object to be attached to the list.
+   */
+  ref?: Ref<HTMLOListElement>
+}
+
+export function OrderedList({children, ...props}: OrderedListProps) {
+  return (
+    <ol className={clsx(styles.List, props.className)} {...props}>
+      <ListContext.Provider value={{variant: ''}}>{children}</ListContext.Provider>
```

Comment:
Why do we need a context provider here? 🤔 context is really expensive and there doesn't seem to be any important stateful logic warranting its usage, or am I missing something? Iterating through children would be faster than using context.

Reply:
So I would have to do the analysis but looping through the objects and rerendering I would think would be more expensive than adding a context object that doesn't get updated. To do it without using either you would need a name-spaced component for each item. I didn't namespace because we can use a single ListItem component which is easier for developers.

Reply:
There's already [data available ](https://blog.logrocket.com/pitfalls-of-overusing-react-context/) on this. I think the larger point here is that I can't see why you need this as you're just passing props, and you shouldn't need context for that? If you're concerned about performance specifically, surely using namespaced component is the fastest option as you don't need to iterate through children or use context.

Reply:
> adding a context object that doesn't get updated

☝️ if it doesn't get updated, why do you need to create a context to store data? 

Would the easiest option not be to just import ListItem as a private component, map the children and forward the variant to it?

Reply:
Mapping over the children would be an expensive operation as well we could have the user include props. But thats an extra step for them. I think mapping over children would be more expensive than adding a context object that wouldn't change.

Reply:
How would we go about testing which approach is more performant?

Reply:
We can discuss the time complexity around this component, but bear in mind that this is going to be used on marketing pages, where we are intentional about not using large data sets. We don't necessarily need to worry about time complexity as much as introducing space complexity for no immediately beneficial reason. 

[Official context docs](https://reactjs.org/docs/context.html#before-you-use-context) state the typical use-case for it, and to avoid using it for situations that don't need it. I can't see any prop-drilling happening, or parent to grandchild data flow, or the parent needing to hold and mutate unique state for each child instance. 

>  context object that wouldn't change

I think it would here though,  because I can't see any memoization being used so you would incur extra rerenders using this approach right now. In fact, you're calling useContext `n` times on each re-render. This makes it no different to iterating through children, except with that approach you're not increasing memory utilisation. 

Going back to the wider point here around this implementation. It feels like we're overengineering this by introducing unnecessary complexity. The user should have a convenient API for something that they expect will be fairly straight forward to render. As a potential user, I'd find it strange to know that there is internal state being maintained in a component that's just an abstraction of `ul` and `li`. 

If we're going for the most performant and pragmatic option, we should create component-level encapsulation around ListItem, that is local to UnorderedList and OrderedList respectively. Something like...

```js

// UnorderedList.tsx

function Root() {} // private

function Item {} // private. use shared styles but otherwise just a wrapper around li

export const OrderedList  = Object.assign(Root, { Item })

```

☝️ I think ListItem should generally be local to each UnorderedList and OrderedList. Yes, there's duplication, but these components are only relevant to their parent components so scoping might make sense.

> Mapping over the children would be an expensive operation as well we could have the user include props

Mapping is functionally pure and deterministic. Built-ins like `map` are also browser optimised and native to JS, unlike hooks like context and their providers which are framework-level functions. Not sure if that statement is universally applicable, but would be interesting to see the benchmarks.

Reply:
The official context docs state that you shouldn't use context for 

> context objects that wouldn't change

Because you should pass the props to the component directly in this case. You have mentioned not wanting to do this because the user shouldn't have to supply more than top level props. Even though this is the most performant and expected way in React. (I do agree that not supplying props on the children does make sense).

Context change would need to happen by passing a new value to the context which this wouldn't do and has no way of doing internally. We could additionally memoize it for further security but at this moment it is the same result.

We could create a component for each list but this would expand our lists to UnorderedList, CheckedList, and OrderedList. If I were using these components I would be less interested in having to import a new list each time I want a slightly different leadingVisual (But maybe our engineers would prefer this approach). Because there are two leading visual options for UnorderedList you will need to manage state anyway to supply these within one component. This comes down to how we would do that.

Since that is the case I think that the only way to know would be to benchmark the two approaches. We can take the time to do that but seeing as I am out starting tomorrow and we wanted to get this done by end of year I wasn't sure if moving it forward was more pressing. Since I am a fan of the context API approach (if it works for you) this could be a good case for passion week to benchmark to find out for sure which makes sense. We could even create an ADR around this based on results because I believe other engineers are making these same decisions.

Reply:
> We could create a component for each list but this would expand our lists to UnorderedList, CheckedList, and OrderedList.

Only UnorderedList and OrderedList are needed. Your switch logic for the icon in ListItem can move to the UnorderedList component file as it doesn't concern OrderedList at all, right? (given you're relying on the default value and returning an empty fragment). 

> Because there are two leading visual options for UnorderedList you will need to manage state anyway to supply these within one component.

No so I'm saying that there is no need for local (useState) or external (useContext) component state at all for this type of component 😅. You can just forward the prop for variant onto the child list item (see example in thread above ☝️). 

> Since I am a fan of the context API approach

I like Context for state management too 😄. I'm trying to be mindful of over-utilising React-specific features when there are already native JS built-in that can achieve the same thing without all the indirection. React provides a nice first-class API for iterating children (React.Children), which is why we use it so heavily in both Primer Brand and in Primer React Components.

> I believe other engineers are making these same decisions.

That's interesting, can you point me at these examples please? AFAIK, the use-cases for Context to date have been to use it to send downstream data back _up_ the tree, or avoid prop-drilling. Otherwise, we use React.Children API to iterate immediate children and forward props, which is already documented in the ADR I linked previously. Has that changed in PRC now?

Reply:
> No so I'm saying that there is no need for local (useState) or external (useContext) component state at all for this type of component 😅. You can just forward the prop for variant onto the child list item (see example in thread above ☝️).

You're right I think my use of "state" was misused in the context of react. There is data to be managed regardless of  the approach we would need to toggle on the leadingVisual depending on the type of Unordered list used. (The data  is what I was referring to as "state")

> I like Context for state management too 😄. I'm trying to be mindful of over-utilising React-specific features when there are already native JS built-in that can achieve the same thing without all the indirection. React provides a nice first-class API for iterating children (React.Children), which is why we use it so heavily in both Primer Brand and in Primer React Components.

I understand the mindfulness here. We are trying to create an anti-pattern within react by avoiding passing props directly to the children. I think that both approaches are functional I just want to know what is more performant. Whichever is more performant has my vote for sure.

> That's interesting, can you point me at these examples please? AFAIK, the use-cases for Context to date have been to use it to send downstream data back up the tree, or avoid prop-drilling. Otherwise, we use React.Children API to iterate immediate children and forward props, which is already documented in the ADR I linked previously. Has that changed in PRC now?

For TreeView it seems that it is being used to pass props to children without manually applying them:
https://github.com/primer/react/blob/main/src/TreeView/TreeView.tsx#L506